awslabs/aws-iot-device-client
# HTTP proxy support

Previous approved staging repo PR: https://github.com/awslabs/private-aws-iot-device-client-staging/pull/4

### Motivation
- Please give a brief description for the background of this change.
Adding HTTP proxy support with basic username and password authentication scheme.
* Introduced new config type: HTTP proxy config. Default path is `~/.aws-iot-device-client/http-proxy.conf`
* Introduced new CLI argument `--http-proxy-config` for overriding default config setting in path `~/.aws-iot-device-client/http-proxy.conf`


### Modifications
* Added `PlainConfig::HttpProxyConfig` with HTTP proxy related parameters
* Implemented several functions from `PlainConfig` to support loading information from CLI, json config then validate.
* Modified `Config::ParseConfigFile` to validate file path and permissions for HTTP proxy config file
* Added `ProxyUtils` for HTTP proxy related validation function utilities
* Unit tests

#### Change summary
Please describe what changes are included in this pull request. 

#### Revision diff summary
If there is more than one revision, please explain what has been changed since the last revision.

### Testing
Unit Test:
* Added parameterized test for `ProxyUtils`
* Added more http proxy config loading validation tests for `/config/Config`
Manual testing:
* Completed simple testing validations for pub/sub example, alone with other existing features.
- CI test run result: <link>


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
